File: lab2/imitation_modeling/src/modeling/device.rs
use std::fmt;

#[derive(Clone, Debug)]
pub struct Device {
    current_time: f64,
    required_time: Option<f64>,

    rand: DeviceRand,
    stats: DeviceStats,
}

impl Device {
    pub fn new(rand: DeviceRand) -> Self {
        Device {
            current_time: 0.0,
            required_time: None,
            rand: rand,
            stats: DeviceStats {
                busy_time: 0.0,
                total_time: 0.0,
                processed: 0,
            },
        }
    }

    /// If device is busy, returns remaining work time
    pub fn get_work_time(&self) -> Option<f64> {
        match self.required_time {
            Some(t) => Some(t - self.current_time),
            None => None,
        }
    }

    pub fn run(&mut self, time: f64) {
        match self.required_time {
            Some(t) => {
                self.current_time += time;

                if self.current_time >= t {
                    // stats
                    self.stats.add_processed();
                    self.stats.add_busy_time(t - (self.current_time - time));
                    self.stats.add_wait_time(self.current_time - t);

                    // change state
                    self.current_time = 0.0;
                    self.required_time = None;
                } else {
                    self.stats.add_busy_time(time);
                }
            }
            None => panic!("Device is not busy"),
        }
    }

    /// Panics if device is busy
    pub fn wait(&mut self, time: f64) {
        if self.required_time.is_some() {
            panic!("Device is busy");
        }

        self.stats.add_wait_time(time);
    }

    pub fn process(&mut self) {
        match self.required_time {
            Some(_) => panic!("Device is busy"),
            None => {
                self.required_time = Some(self.rand.next_rand());
                self.current_time = 0.0;
            }
        }
    }

    pub fn get_stats(&self) -> DeviceStats {
        self.stats.clone()
    }
}


#[allow(dead_code)]
#[derive(Clone, Debug)]
pub enum DeviceRand {
    Exponential(f64),
    Uniform(f64, f64),
    Fixed(f64),
    Mixed(Vec<DeviceRand>),
}

impl DeviceRand {
    pub fn next_rand(&self) -> f64 {
        let mut n: f64 = rand::random();
        while n == 0.0f64 {
            n = rand::random();
        }

        match self {
            DeviceRand::Exponential(lambda) => -(1.0 / lambda) * n.ln(),
            DeviceRand::Uniform(a, b) => a + (b - a) * n,
            DeviceRand::Fixed(a) => *a,
            DeviceRand::Mixed(rands) => rands.iter().map(|rand| rand.next_rand()).sum(),
        }
    }
}

#[derive(Clone)]
pub struct DeviceStats {
    busy_time: f64,
    total_time: f64,
    processed: usize,
}

impl DeviceStats {
    pub fn add_busy_time(&mut self, time: f64) {
        self.busy_time += time;
        self.total_time += time;
    }

    pub fn add_wait_time(&mut self, time: f64) {
        self.total_time += time;
    }

    pub fn add_processed(&mut self) {
        self.processed += 1;
    }
}

impl fmt::Debug for DeviceStats {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "DeviceStats {{")?;
        writeln!(f, "   busy_time: {:?}", self.busy_time)?;
        writeln!(f, "   total_time: {:?}", self.total_time)?;
        writeln!(f, "   processed: {:?}", self.processed)?;
        writeln!(f, "   utilization: {:?}", self.busy_time / self.total_time)?;
        writeln!(f, "}}")
    }
}

File: lab4/src/main/java/com/example/modeling/components/device/Device.java
package com.example.modeling.components.device;

import java.util.Optional;

import lombok.EqualsAndHashCode;
import lombok.Getter;

@EqualsAndHashCode
public class Device {
    final DeviceRand rand;
    final Stats stats;
    final String name;

    double current_time;
    Optional<Double> required_time;

    public Device(DeviceRand rand, String name) {
        this.rand = rand;
        this.current_time = 0.0;
        this.required_time = Optional.empty();
        this.stats = new Stats();
        this.name = name;
    }

    public Optional<Double> getWorkTime() {
        return Optional.ofNullable(required_time
                .map(t -> t - this.current_time)
                .orElse(null));
    };

    /*
     * Throws exception if device is not busy
     */
    public boolean run(double time) {
        boolean done = false;

        if (this.required_time.isPresent()) {
            var t = this.required_time.get();

            this.current_time += time;

            if (this.current_time >= t) {
                done = true;

                // stats
                this.stats.addProcessed();
                this.stats.addBusyTime(t - (this.current_time - time));
                this.stats.addWaitTime(this.current_time - t);

                // change state
                this.current_time = 0.0;
                this.required_time = Optional.empty();
            } else {
                this.stats.addBusyTime(time);
            }
        } else {
            throw new IllegalStateException("Device is not busy");
        }

        return done;
    }

    /*
     * Throws exception if device is busy
     */
    public void wait(double time) {
        this.required_time.ifPresentOrElse(t -> {
            throw new IllegalStateException(
                "Device is busy. Required time left: " + (t - this.current_time));
        }, () -> {
            this.stats.addWaitTime(time);
        });
    }

    /*
     * Throws exception if device is busy
     */
    public void process() {
        if (this.required_time.isPresent()) {
            throw new IllegalStateException(
                "Device is busy, Required time left: " + (this.required_time.get() - this.current_time));
        } else {
            this.required_time = Optional.of(this.rand.next_rand());
            this.current_time = 0.0;
        }
    }

    public Stats getStats() {
        return this.stats;
    };

    @Getter
    @EqualsAndHashCode
    public class Stats {
        private double busy_time = 0;
        private double total_time = 0;
        private long processed = 0;

        public void addBusyTime(double time) {
            this.busy_time += time;
            this.total_time += time;
        }

        public void addWaitTime(double time) {
            this.total_time += time;
        }

        public void addProcessed() {
            this.processed += 1;
        }

        public double getUtilization() {
            return this.total_time != 0
                    ? this.busy_time / this.total_time
                    : 0;
        }

        public String toString() {
            String format = "%s:{busy_time=%.2f, total_time=%.2f, utilization=%.2f, processed=%d}";

            return String.format(
                    format,
                    Device.this.name,
                    this.busy_time,
                    this.total_time,
                    this.getUtilization(),
                    this.processed);
        }
    }

    @FunctionalInterface
    public interface DeviceRand {
        double next_rand();
    }
}
